Fix #23868: Game crashes when moving to a corner of my park

QuestionableDeer opened the pull request:
Based on the discussion of this issue, it appears to be caused by an oversight in the 3D sign text cutoff fix (7de61c7) where in certain cases the index can wrap around and cause invalid input to substr. The documentation on what happened in this case was ambiguous, but in this case it leads to an exception. The fix is to have a check for a zero index before decrementing in the line wrapping code.

This one's one me, the documentation for `substr` indicates that when `pos + count` is greater than string length, it will simply return the remainder of the string. However if `count` itself is larger than the string's size by itself, this is an exception. I find that design confusing, but the fix is simple at least.

I've verified it works in the case of a sign with all spaces that was breaking in the discussion thread, but I do still need someone to check in the originally problematic parks, since I don't know if something else in the rendering code has been affected separately.

AaronVanGeffen commented:
@deurklink Could you test this?

deurklink commented:
I gave it a test, the affected park no longer crashes. I can create large 3D signs with 16 or more spaces without the game crashing.

mixiate commented:
There are some differences between RCT2/0.3.5 and this.

RCT2:
![rct2signs](https://github.com/user-attachments/assets/e535f0b4-e5da-42bd-82cb-406a728e66b4)

This:
![signs](https://github.com/user-attachments/assets/06286c4a-6321-4398-8457-d84cfa1f9e45)


[Six Flags Great Adventure.zip](https://github.com/user-attachments/files/18915353/Six.Flags.Great.Adventure.zip)

QuestionableDeer commented:
Looks like the differences are mostly in the back row, is that right?

mixiate commented:
Yeah. Ultimately I think the underlying issue here is that [4288589](https://github.com/OpenRCT2/OpenRCT2/commit/4288589f11538ddf6b314b5d4872a46719f5f11c) changed how it worked too much. [I had a go at re writing it equivalent to how it used to work](https://github.com/OpenRCT2/OpenRCT2/compare/develop...mixiate:OpenRCT2:fix-3d-signs-text-layout).

QuestionableDeer commented:
> Yeah. Ultimately I think the underlying issue here is that [4288589](https://github.com/OpenRCT2/OpenRCT2/commit/4288589f11538ddf6b314b5d4872a46719f5f11c) changed how it worked too much. [I had a go at re writing it equivalent to how it used to work](https://github.com/OpenRCT2/OpenRCT2/compare/develop...mixiate:OpenRCT2:fix-3d-signs-text-layout).

Does that restore the sign behavior? If it does, I'm inclined to simply things and go with it, since the current structure is running into more and more special cases each iteration.

QuestionableDeer commented:
> As far as I can tell it does. It handles the signs in the pictures and a couple of other I tried to break it with, but i'm a little worried there's some edge case sign it doesn't do properly. I had to change the original 0.3.5 code a little to fit it in with the iterators and views.

I tried copying that portion of the file to my branch to check and in my case I'm seeing it cut off the first letter of the second line? So "Meow Meow" becomes "Meow eow". This is just plopping down a new sign in the first scenario I click on. Are there any other changes you made outside that function?

mixiate commented:
Ah you're right, I pushed an old commit. Sorry.

[This should be working](https://github.com/OpenRCT2/OpenRCT2/compare/develop...mixiate:OpenRCT2:fix-3d-signs-text-layout).

QuestionableDeer commented:
Closing this in preference to #23870 which addresses the issues more fundamentally with a refactor to clean up additional changed behavior from base RCT2.
